### PR TITLE
Change to use trunk-based development model instead of GitFLow

### PR DESCRIPTION
**Pull request by MarkoSagadin**
## Description

Change to use trunk-based development model instead of GitFLow.

Related: #31

## Notes for reviewer

Please check if the new description is clear. Many word were written about TBD (https://trunkbaseddevelopment.com/), and I did not want to repeat them here. Also, if you think that we should add something that we care about but is not written, please comment.

### CODE REVIEW

**Comment by MarkoSagadin** [comment]
Current dependencies on/for this PR:
* dev
  * **PR #32** <a href="https://app.graphite.dev/github/pr/IRNAS/irnas-guidelines-docs/32" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/IRNAS/irnas-guidelines-docs/32?utm_source=stack-comment).

File: docs/github_projects_guidelines.md [hunk 1]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
-  committed to it in form of a GitHub Pull Request. Nothing should be committed
-  to it directly (exceptions can be made for GitHub CI workflows, which can
-  otherwise misbehave).
-- `dev` branch contains the latest development changes for the next release.
-  Commits to it can be done to it directly or in form of GitHub Pull requests.
+Each repository has a single long-lived, default, branch named `main`.
 
-Besides these two branches, various short-lived branches exist in this workflow.
-They are classified into three categories: feature, release, and hotfix.
+There are two rules that should be followed when adding new code to the `main`:
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:61]:
```suggestion
There are two rules that should be followed when adding new code to `main`:
```

File: docs/github_projects_guidelines.md [hunk 2]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
-- release branches are used for the preparation of a new release. They allow for
-  last-minute changes. They are always created from the `dev` branch and merged
-  into the `master` branch. They are named as `release/<version>`, where
-  `<version>` indicates the version of the next release.
-- hotfix branches are used to address unplanned changes to the release in
-  production. The essence is that the work of team members can continue, while
-  another person is preparing a quick production fix. Hotfix branches must
-  always branch from the `master` branch and merge back into it.
+- It should **never break the build** (whatever _build_ means for a specific
+  repository). Bugs should be caught by the CI or by code reviews.
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:64]:
```suggestion
- It should **never break the build** (whatever _build_ means for a specific
  repository). Bugs should be caught by CI or by code reviews.
```

File: docs/github_projects_guidelines.md [hunk 3]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
+- `main` should **always be release ready**, even if there is work in progress.
+  Functionalities that might need some more development time to become useful
+  should not be included in the build until they are ready.
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:67]: Maybe add some examples of how to achieve this.

For example, by either not merging a feature branch until it is ready. Not including unfinished files in CMake, ...

File: docs/github_projects_guidelines.md [hunk 4]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
 
-It is important that after every merge into the `master` branch a Pull Request
-is made from `master` back into `dev`. This brings in the changes that were made
-during the release process in the release and hotfix branches.
+Those two rules can sometimes be hard to follow in practice due to many internal
+and external factors. If a deviation from those two rules happens effort should
+be made to bring the `main` branch into the correct state.
 
-### Deviations from GitFlow article
+New code can be added to the `main` in two ways:
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:73]:
```suggestion
New code can be added to `main` in two ways:
```

File: docs/github_projects_guidelines.md [hunk 5]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
 
-### Feature branches
+#### Feature branches
 
 A few things should be kept in mind when working with feature branches:
 
-- They should be focused on accomplishing one thing.
-- They should be small, they should take up to 1 day of work to complete and to
-  open Pull Request back into the `dev` branch.
+- They should be focused on accomplishing one small thing.
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:82]: What is **small**? How do you quantify it?

File: docs/github_projects_guidelines.md [hunk 6]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
+- They should take up to 1 day of work to complete and to open Pull Request back
+  into the `main` branch for the code review.
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:84]:
```suggestion
- They should take up to 1 day of work to complete.
- Once complete, a Pull Request into `main` should be created for the code review.
```

About the "1 day of work" thing. Maybe add some text here explaining that larger work must be split into smaller tasks, that are planned to take up to 1 day of work.
Also, there are some that will take more than 1 day, and can not further be split (like bug-hunts for example) - how do we handle those?

File: docs/github_projects_guidelines.md [hunk 7]
```diff
@@ -323,9 +314,8 @@ a project and should be always kept in sync.
 
 ### Creating releases
 
-Since we are following the [GitFlow] branching model, a GitHub release is always
-created after a release branch is merged into the master branch and the tag has
-been created.
+Since we are following the [trunk-based development] branching model, a GitHub
+release is created directly from a commit that was tagged with release tag.
```
Comment (TjazVracko) [inline docs/github_projects_guidelines.md:318]:
```suggestion
Since we are following the [trunk-based development] branching model, a GitHub
release is created directly from a commit that was tagged with the release tag.
```

**Review by TjazVracko — CHANGES_REQUESTED** [review]
Good stuff.

I think we also need a section about code-review and the branch merging process.

You mention a feature branch requiring a review to be merged here. Maybe link to a separate code-review section there.
We should also discuss if all feature branches should be squashed into a single commit before being merged.
Maybe all features should be a single commit in the first place (like Zephyr does). What are your thoughts?

File: docs/github_projects_guidelines.md [hunk 8]
```diff
@@ -44,68 +44,59 @@ working with projects hosted on GitHub.
...
+- `main` should **always be release ready**, even if there is work in progress.
+  Functionalities that might need some more development time to become useful
+  should not be included in the build until they are ready.
 
-It is important that after every merge into the `master` branch a Pull Request
-is made from `master` back into `dev`. This brings in the changes that were made
-during the release process in the release and hotfix branches.
+Those two rules can sometimes be hard to follow in practice due to many internal
+and external factors. If a deviation from those two rules happens effort should
+be made to bring the `main` branch into the correct state.
```
[the remark this answers is not in the document]
Comment (MarkoSagadin) [reply docs/github_projects_guidelines.md:71]: Yep, that is better, applied the commit.

File: docs/github_projects_guidelines.md (hunk 3 above)
Comment (MarkoSagadin) [reply docs/github_projects_guidelines.md:67]: Added.

**Comment by MarkoSagadin** [comment]
@TjazVracko you raised several good points. Sections about that describe our development process (PR process, code reviews, splitting work into smaller PRs, merging choice) should be written. I will open an issue that will address them.

File: CHANGELOG.md [hunk 9]
```diff
@@ -24,6 +24,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
     were moved to `docs/github_project_guidelines.md`.
 -   Various documents related to tooling, such as clang-format and gitlint, are
     now in `tools` directory.
+-   Branching model to the Trunk-based development from GitFlow.
```
Comment (TjazVracko) [inline CHANGELOG.md:27]:
```suggestion
-   Branching model to the Trunk-based development model from GitFlow.
```

File: docs/developer_guidelines.md [hunk 10]
```diff
@@ -33,6 +38,76 @@
...
+
+- It should **never break the build** (whatever _build_ means for a specific
+  repository). Bugs should be caught by automated tests, static analysis tools
+  or by code reviews.
+- `main` should **always be release-ready**, even if there is work in progress.
+
+Functionalities that might need some more development time to become useful
+should not be included in the build until they are ready. For example: don't
+include unfinished source files in the `CMakelists.txt` files or don't merge
+feature branches until they are ready, etc.
```
Comment (TjazVracko) [inline docs/developer_guidelines.md:68]:
```suggestion
include unfinished source files in `CMakelists.txt` files, don't merge
feature branches until they are ready, etc.
```

File: docs/developer_guidelines.md [hunk 11]
```diff
@@ -33,6 +38,76 @@
...
+circumstances. If a deviation from those two rules happens effort should be made
+to bring the `main` branch into the correct state (**buildable** and
+**release-ready**).
+
+New code can be added to `main` in two ways:
+
+- either by committing to it directly, or
+- in the form of feature branches.
+
+#### Committing directly to the `main`
```
Comment (TjazVracko) [inline docs/developer_guidelines.md:80]:
```suggestion
#### Committing directly to `main`
```

File: docs/developer_guidelines.md [hunk 12]
```diff
@@ -33,6 +38,76 @@
...
+commit directly to the `main` branch.
+
+As soon you are:
+
+- working in a repository alongside other developers,
+- want your work reviewed or
+- making critical changes
+
+then you probably want to create a feature branch for your changes and have your
+work reviewed by a another developer.
```
Comment (TjazVracko) [inline docs/developer_guidelines.md:92]:
```suggestion
work reviewed by another developer.
```

File: docs/developer_guidelines.md [hunk 13]
```diff
@@ -33,6 +38,76 @@
...
+
+#### Feature branches
+
+A few things should be kept in mind when working with the feature branches:
+
+- They should be focused on accomplishing one small thing.
+- They should take up to 1 day of work to complete and to open Pull Request back
+  into the `main` branch for the code review.
```
Comment (TjazVracko) [inline docs/developer_guidelines.md:100]: You did not use my earlier suggestion. I am repeating it here.

```suggestion
- They should take up to 1 day of work to complete.
- Once complete, a Pull Request into `main` should be created for the code review.
```

File: docs/developer_guidelines.md [hunk 14]
```diff
@@ -33,6 +38,76 @@
...
+
+The rapid, small commits of trunk-based development make code review a more
+efficient process. With small branches, developers can quickly see and review
+small changes. This is far easier compared to a long-lived feature branch where
+a reviewer reads pages of code or manually inspects a large surface area of code
+changes.
+
+Use `feature/<branch name>` format when naming feature branches, for example
```
Comment (TjazVracko) [inline docs/developer_guidelines.md:108]:
```suggestion
Use the `feature/<branch name>` format when naming feature branches, for example
```

**Review by TjazVracko — CHANGES_REQUESTED** [review]
I agree with all of the content :) 
I have marked some grammatical mistakes, otherwise, this is good :+1: 
Please run the whole document through Grammarly before publishing.

**Comment by MarkoSagadin** [comment]
@TjazVracko I have already created a cleanup PR in the stack of the PRs, in which I have run the markdown files through the Grammarly. You suggestions were either already applied or I just manually applied them right now.

**Comment by MarkoSagadin** [comment]
@MarkoSagadin merged this pull request with [Graphite](https://app.graphite.dev/github/pr/IRNAS/irnas-guidelines-docs/32).
